### ISSUE TITLE
Unit tests

### ISSUE THREAD

**Issue by bminixhofer** [bminixhofer/tractjs#9]
As mentioned in https://github.com/bminixhofer/tractjs/pull/7 unit tests would be very nice. At the moment this is blocked by #8 but I'll see if I can fix that soon.

I'm open to the choice of framework, I've used Mocha in the past but Jest also looks good to me.

cc @danielbank if you're still interested.

**Comment by danielbank** [comment]
Yes I've been looking at this and trying to setup a test with Jest in the project

**Comment by bminixhofer** [comment]
Great, let me know if you run into any problems!

**Comment by danielbank** [comment]
I have a start on it here: https://github.com/bminixhofer/tractjs/compare/master...danielbank:unit-tests?expand=1#diff-00b8729173de8959e19581ff4b8039cc

I'm have a really hard time trying to load the WebWorker in a test.  Currently dealing with:
```
  console.error
    SyntaxError: Cannot use import statement outside a module
        at Function (<anonymous>)
        at /Users/DanielBank/workspace/tractjs/wrapper/node_modules/jsdom-worker/src/index.js:89:18
```

**Comment by bminixhofer** [comment]
Hi, sorry for the delay. And great starting point!

You're trying to load `src/worker.ts` directly. It has to be built with Rollup before using it somewhere. The problem is that at the moment the worker is always inlined (and as mentioned in #8 I believe it has some web-specific code). 

I'm working on generating a separate, Node.js ready worker now for the CommonJS build.

**Comment by danielbank** [comment]
What you're saying is making more sense as I work with it.  Though it does raise a question in my mind:

Node.js support can be added and that's all well and good (e.g. someone might want to incorporate it into a ExpressJS server).  But shouldn't the behavior of the Web Workers (Node.js and Browser) be covered by a unit test?  I was looking at [jsdom](https://github.com/jsdom/jsdom) and specifically [jsdom-worker](https://www.npmjs.com/package/jsdom-worker) as a means of at least simulating the Browser environment for the web worker code even though it is running in Jest on top of Node.

I also have been delayed in working on this, but should be back at it again this weekend.

**Comment by bminixhofer** [comment]
That's a good point. If tractjs has significantly different code in Node.js and the browser, unit tests for both would be needed. 

JSDom looks good and might be one solution. Another one would be using [web-worker](https://github.com/developit/web-worker) to avoid the differences between Node.js workers and web workers altogether. But I couldn't get that to work nicely with Rollup yet.

**Comment by bminixhofer** [comment]
Good news! I made great progress towards Node.js support at https://github.com/bminixhofer/tractjs/tree/nodejs-support. I am using [web-worker](https://github.com/developit/web-worker/) now so there are no differences in web worker code between Node and the browser. Once the tests pass I'll merge into master.

The only difference is in loading the model file. The browser bundle uses `fetch`, while the node bundle uses `fs.readFile`. I solved this by adding two utility modules with the same API that the bundler can switch between (at [wrapper/src/utils](https://github.com/bminixhofer/tractjs/tree/nodejs-support/wrapper/src/utils)). 

Now if we run the tests in Node.js the only untested code is [wrapper/src/utils/browser.ts](https://github.com/bminixhofer/tractjs/blob/nodejs-support/wrapper/src/utils/browser.ts) which given the extra effort is not worth it to test in my opinion.

There could of course also be some issue with the external packages like `web-worker` that break tractjs but since we have integration tests anyway that run in an actual Firefox release I think we're good in that regard.

Oh and I'll also add a `CONTRIBUTING.md` file with a short intro to the build process to make it easier for contributors in the future (since I noticed that all the build commands in package.json like `build-wrapper`, `build-pkg` etc. could be confusing).

**Comment by bminixhofer** [comment]
FYI, this is now merged into master and all the integration tests are passing. I also added a [CONTRIBUTING.md](https://github.com/bminixhofer/tractjs/blob/master/CONTRIBUTING.md) with some info about the build process.

@danielbank your code from https://github.com/danielbank/tract-js/tree/unit-tests should now work if you use the worker built to `wrapper/dist/worker.js`. And since there are no differences in code between browser and Node workers going ahead with the test structure you were using looks good to me!

**Comment by danielbank** [comment]
Very slow going work for me on this.  I am now able to load the built `./dist/worker.js` into my unit test:

```
import worker from '../dist/worker';
```

However this is an empty object (it doesn't export anything), so I can't call `postMessage()` or `addEventListener()` on it.

It seems when you use it in `index.ts`, you are calling:

```
import worker_data from "../dist/worker.js";
const worker = new Worker(worker_data);
```

This seems to just be used by the rollup script to inline it in the built `tractjs.js`.  So I don't think I can test the internals this way.

My current work is here: https://github.com/bminixhofer/tractjs/compare/master...danielbank:unit-tests?expand=1

**Comment by bminixhofer** [comment]
All that Rollup setup is kind of complex unfortunately :(

BUT: You are actually almost there. The line:

```javascript
import worker_data from "../dist/worker.js";
```

reads the worker as data URI instead of directly importing it because of this line in `wrapper/rollup.js`:
```javascript
// ...
      url({
        include: "./dist/worker.js",
        limit: Infinity, // always inline the worker
      }),
// ...
```

so the problem is that you directly import the worker which doesn't leave you with anything since it is just an IIFE that does  various calls to `addEventListener`.

If you change the code to:

```javascript
import Worker from "web-worker";

const worker = new Worker("./dist/worker.js"); // if the filepath doesn't work, try a data URI e. g. with https://www.npmjs.com/package/datauri
```

you should get a fully functional worker which you can call `addEventListener`, `postMessage` etc. on!

**Comment by danielbank** [comment]
Okay, I got the Worker API working inside my test.  One problem down 😉 

One issue I am having is that I need to install [@types/jest](https://www.npmjs.com/package/@types/jest) in order to clear TypeScript errors in the test files.  However when I include this devDependency, rollup is trying to bundle the test types and it breaks the build when it gets to the dts-bundle-generator step with an error:

```
Compiling input files...
../node_modules/@types/mocha/index.d.ts(2477,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'beforeEach' must be of type 'Lifecycle', but here has type 'HookFunction'.
```

**Comment by bminixhofer** [comment]
Great! I'll have a luck at the .d.ts bundle generator problem. I'm not that happy with having to use that in general (it's only needed for correctly generating docs).

**Comment by danielbank** [comment]
I'm still working on this.  I resolved the issue with Jest types and the .d.ts bundle generator, I just needed to add `"types":"jest"]` to the `tsconfig.json`.  Right now I am working on keeping the tests from being bundled into the `dist` folder.  Hopefully I'll be working on actual tests by next weekend.

One thing I have thinking about:

I can test the worker code by loading into the test script with:
```
worker = new Worker(await datauri("./dist/worker.js") as string);
```
However these tests do not show up in the Istanbul test coverage (which shows which lines of code are being executed).  As far as I can tell, Istanbul cannot test code which is dynamically loaded like this.  On the other hand, I can write tests against the generated `tractjs.js` file and it does show up in the coverage report.

Maybe there is some merit to testing the worker code (even though it doesn't show up in coverage report).  This would actually be testing the internals of the black box.  However, testing the `tractjs.js` file could be more important because that is testing the library's API and ensuring consistent behavior (at least for Node).  I wasn't able to write tests for `tractjs.esm.js` because the tests run in a Node.  As you said before, the only difference is how the `load()` works and there is no real benefit to testing that code.

**Comment by bminixhofer** [comment]
Great! And yes its probably near impossible to detect code being loaded this way. 

Thanks for sharing your thoughts on this, and thanks for your dedication to keep working on it, even though there's quite a few issues.

I agree with your assessment that chasing code coverage can be counterproductive; there's also some interesting [blog posts](https://itnext.io/the-myth-of-100-code-coverage-c7d4c789700d) on that topic.

Along those lines I'm thinking that testing the public API (i. e. the `tractjs.js` file) should have the highest priority. Particularly the non-happy paths. We might discover more issues like fixed in https://github.com/bminixhofer/tractjs/pull/7 this way.

Testing the internal worker code could be useful (and it is definitely useful that we at least have a way to do so now), but should have a lower priority, and after thinking about this a bit I am not certain that it would be good (more code, more coupling). But I am certain that unit tests for the public API are needed.
